File: sdks/php/src/Webhook.php
<?php

declare(strict_types=1);

namespace HookSniff;

/**
 * HookSniff Webhook Signature Verification
 *
 * Verifies incoming webhook signatures using HMAC-SHA256.
 * Compatible with Standard Webhooks format (whsec_ prefix secrets).
 *
 * Usage:
 *   $payload = Webhook::verify('whsec_...', $rawBody, $headers);
 */
class Webhook
{
    private const TIMESTAMP_TOLERANCE_SECONDS = 5 * 60; // 5 minutes

    private string $secret;

    /**
     * Create a new Webhook verifier.
     *
     * @param string $secret The endpoint's signing secret (e.g., "whsec_base64encoded...")
     */
    public function __construct(string $secret)
    {
        $this->secret = self::decodeSecret($secret);
    }

    /**
     * Verify a webhook payload against its signature headers.
     *
     * @param string $payload The raw request body
     * @param array<string, string> $headers The request headers
     * @return mixed The parsed payload if verification succeeds
     * @throws WebhookVerificationError if verification fails
     */
    public function verify(string $payload, array $headers): mixed
    {
        // Normalize headers to lowercase
        $normalized = [];
        foreach ($headers as $key => $value) {
            $normalized[strtolower($key)] = $value;
        }

        // Support both svix- and webhook- prefixed headers
        $msgId = $normalized['svix-id'] ?? $normalized['webhook-id'] ?? null;
        $timestamp = $normalized['svix-timestamp'] ?? $normalized['webhook-timestamp'] ?? null;
        $signature = $normalized['svix-signature'] ?? $normalized['webhook-signature'] ?? null;

        if ($msgId === null) {
            throw new WebhookVerificationError('Missing webhook-id header');
        }
        if ($timestamp === null) {
            throw new WebhookVerificationError('Missing webhook-timestamp header');
        }
        if ($signature === null) {
            throw new WebhookVerificationError('Missing webhook-signature header');
        }

        // Validate timestamp (prevent replay attacks)
        $timestampNum = (int) $timestamp;
        if ($timestampNum === 0 && $timestamp !== '0') {
            throw new WebhookVerificationError('Invalid webhook-timestamp header');
        }

        $now = (int) time();
        if (abs($now - $timestampNum) > self::TIMESTAMP_TOLERANCE_SECONDS) {
            throw new WebhookVerificationError(
                sprintf('Webhook timestamp is too old or too new (tolerance: %ds)', self::TIMESTAMP_TOLERANCE_SECONDS)
            );
        }

        // Compute expected signature
        $content = $msgId . '.' . $timestamp . '.' . $payload;
        $expectedSig = base64_encode(hash_hmac('sha256', $content, $this->secret, true));
        $expected = 'v1,' . $expectedSig;

        // Timing-safe comparison
        if (!self::verifySignature($expected, $signature)) {
            throw new WebhookVerificationError('Invalid webhook signature');
        }

        // Parse and return payload
        $decoded = json_decode($payload, true);
        return $decoded !== null ? $decoded : $payload;
    }

    /**
     * Sign a payload (for testing or server-side webhook sending).
     *
     * @param string $msgId The message ID
     * @param int $timestamp Unix timestamp
     * @param string $payload The payload to sign
     * @return string The signature string (e.g., "v1,base64hmac")
     */
    public function sign(string $msgId, int $timestamp, string $payload): string
    {
        $content = $msgId . '.' . $timestamp . '.' . $payload;
        $hmac = base64_encode(hash_hmac('sha256', $content, $this->secret, true));
        return 'v1,' . $hmac;
    }

    /**
     * Static convenience method for one-shot verification.
     *
     * @param string $secret The signing secret
     * @param string $payload The raw request body
     * @param array<string, string> $headers The request headers
     * @return mixed The parsed payload
     * @throws WebhookVerificationError
     */
    public static function verifyWithSecret(string $secret, string $payload, array $headers): mixed
    {
        $wh = new self($secret);
        return $wh->verify($payload, $headers);
    }

    /**
     * Decode a whsec_ prefixed secret to raw bytes.
     */
    private static function decodeSecret(string $secret): string
    {
        // Strip whsec_ prefix if present
        $raw = str_starts_with($secret, 'whsec_') ? substr($secret, 6) : $secret;

        // Try base64 decode
        $decoded = base64_decode($raw, true);
        return $decoded !== false ? $decoded : $raw;
    }

    /**
     * Verify that a signature matches using timing-safe comparison.
     */
    private static function verifySignature(string $expected, string $actual): bool
    {
        // Each signature can be comma-separated (v1 sig1, v1 sig2, ...)
        $signatures = array_map('trim', explode(',', $actual));

        foreach ($signatures as $sig) {
            // Strip version prefix
            $parts = explode(',', $sig, 2);
            $signaturePart = count($parts) > 1 ? $parts[1] : $parts[0];

            // Strip version prefix from expected too
            $expectedParts = explode(',', $expected, 2);
            $expectedSig = count($expectedParts) > 1 ? $expectedParts[1] : $expectedParts[0];

            // Length check before timing-safe compare
            if (strlen($expectedSig) !== strlen($signaturePart)) {
                continue;
            }

            if (hash_equals($expectedSig, $signaturePart)) {
                return true;
            }
        }

        return false;
    }
}

File: sdks/go/webhook.go
package hooksniff

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

const timestampTolerance = 5 * time.Minute

var (
	ErrMissingWebhookID        = errors.New("hooksniff: missing webhook-id header")
	ErrMissingWebhookTimestamp = errors.New("hooksniff: missing webhook-timestamp header")
	ErrMissingWebhookSignature = errors.New("hooksniff: missing webhook-signature header")
	ErrInvalidTimestamp        = errors.New("hooksniff: invalid webhook-timestamp header")
	ErrTimestampTooOld         = errors.New("hooksniff: webhook timestamp is too old or too new")
	ErrInvalidSignature        = errors.New("hooksniff: invalid webhook signature")
)

// Webhook verifies incoming webhook signatures.
type Webhook struct {
	secret []byte
}

// NewWebhook creates a new Webhook verifier.
// The secret should be the endpoint's signing secret (e.g., "whsec_base64encoded...").
func NewWebhook(secret string) *Webhook {
	return &Webhook{secret: decodeSecret(secret)}
}

// decodeSecret strips the whsec_ prefix and base64-decodes the secret.
func decodeSecret(secret string) []byte {
	raw := secret
	if strings.HasPrefix(raw, "whsec_") {
		raw = raw[6:]
	}

	decoded, err := base64.StdEncoding.DecodeString(raw)
	if err != nil {
		return []byte(raw)
	}
	return decoded
}

// Verify verifies a webhook payload against its signature headers.
// Returns the parsed payload as a map if verification succeeds.
func (w *Webhook) Verify(payload []byte, headers map[string]string) (map[string]interface{}, error) {
	// Normalize headers to lowercase
	normalized := make(map[string]string)
	for k, v := range headers {
		normalized[strings.ToLower(k)] = v
	}

	// Support both svix- and webhook- prefixed headers
	msgID := normalized["svix-id"]
	if msgID == "" {
		msgID = normalized["webhook-id"]
	}
	timestamp := normalized["svix-timestamp"]
	if timestamp == "" {
		timestamp = normalized["webhook-timestamp"]
	}
	signature := normalized["svix-signature"]
	if signature == "" {
		signature = normalized["webhook-signature"]
	}

	if msgID == "" {
		return nil, ErrMissingWebhookID
	}
	if timestamp == "" {
		return nil, ErrMissingWebhookTimestamp
	}
	if signature == "" {
		return nil, ErrMissingWebhookSignature
	}

	// Validate timestamp
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return nil, ErrInvalidTimestamp
	}

	now := time.Now().Unix()
	if abs(now-ts) > int64(timestampTolerance.Seconds()) {
		return nil, ErrTimestampTooOld
	}

	// Compute expected signature
	content := fmt.Sprintf("%s.%s.%s", msgID, timestamp, string(payload))
	mac := hmac.New(sha256.New, w.secret)
	mac.Write([]byte(content))
	expectedSig := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	expected := "v1," + expectedSig

	// Verify signature (supports multiple comma-separated signatures)
	if !verifySignature(expected, signature) {
		return nil, ErrInvalidSignature
	}

	// Parse JSON payload
	var result map[string]interface{}
	if err := json.Unmarshal(payload, &result); err != nil {
		// Return raw payload as string if not JSON
		return map[string]interface{}{"_raw": string(payload)}, nil
	}

	return result, nil
}

// Sign signs a payload for testing or server-side webhook sending.
func (w *Webhook) Sign(msgID string, timestamp time.Time, payload []byte) string {
	ts := strconv.FormatInt(timestamp.Unix(), 10)
	content := fmt.Sprintf("%s.%s.%s", msgID, ts, string(payload))
	mac := hmac.New(sha256.New, w.secret)
	mac.Write([]byte(content))
	sig := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	return "v1," + sig
}

// verifySignature checks if any of the comma-separated signatures match.
func verifySignature(expected, actual string) bool {
	signatures := strings.Split(actual, ",")
	for _, sig := range signatures {
		sig = strings.TrimSpace(sig)

		// Strip version prefix
		parts := strings.SplitN(sig, ",", 2)
		sigPart := sig
		if len(parts) > 1 {
			sigPart = parts[1]
		}

		expectedParts := strings.SplitN(expected, ",", 2)
		expectedPart := expected
		if len(expectedParts) > 1 {
			expectedPart = expectedParts[1]
		}

		if len(expectedPart) != len(sigPart) {
			continue
		}

		if hmac.Equal([]byte(expectedPart), []byte(sigPart)) {
			return true
		}
	}
	return false
}

func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}
